1077 persistent anchors

jmdelfa opened the pull request:
* **Tickets addressed:** Closes #1077 & Closes #1205
* **Review:** By commit  <!-- Choose from: "by commit", "by file" -->
* **Merge strategy:** Merge (no squash)  <!-- Choose from: "merge (no squash)", "squash and merge" -->

## Description
Inclusion of Persistent Anchors in Coexistence Goal. The anchor definition is optional so that backwards compatibility with older Coexistence Goals is guaranteed. An anchor can be either disabled, to the start of a directive or to its end.
Implementation has focused on mitigating the impact on existing code trying to reuse as much as possible existing AERIE infrastructure. Even so, 40 classes haven been added/modified/deleted. TimeExpression has been heavily re-architected as there were many legacy relations. CoexistenceGoal and PrioritySolver have been modified to account for the indentification and resolution of conflicts related to anchors. The need for new conflict types was explored and discarded for the sake of simplicity.

## Verification
Multiple tests have been added, including a dedicated TestPersistentAnchors that explores all possible combinations of anchors and existing/non-existing directives in the plan.

## Documentation
No documentation is invalidated. New documentation needs to be produced to explain the anchor behaviour

## Future work
This work is self contained and complete. Future extensions such as deeper control on when/how anchors are created should derive from user experience

jmdelfa committed:
Re-engineering TimeExpression. Unused/obsolete classes have been removed. All time expressions are actually relative. TimeExpressoinRelatveFixed replaced refactored as TimeExpressionRelativeSimple

jmdelfa committed:
Inclusion of BidiMap to query relation between SchedulingActivityDirectiveIds and ActivityDirectiveIds

jmdelfa committed:
Modified SchedulingActivityDirective providing updated of and copyOf methods. Also add a type method returning the AcitivityType of the directive. Changes used in PrioritySolver, MissingActivityTemplateConflict and goals

jmdelfa committed:
Adding EDSL syntax for PersistentTimeAnchor

jmdelfa committed:
Adding PersistentTimeAnchor java data structures and mechanisms to add/update anchorIds
PersistentTimeAnchor.java adds the three states of an anchor: Disabled, anchoring to start or anchoring to end
ActivityExpression.java adds a new matches method to evaluate the time offset considering the absolute time, which is calculated by iteratively navigating the anchors.
GraphQLMerlinService.java provides a new method to update the anchorids in a list of activities
MerlinService.java adds a new method to update a list of SchedulingActivityDirectives from a plan, replacing the anchorids generated by the scheduler by new anchorids provided by the database
Plan.java and PlanInMemory.java provide methods to replace an old activity by a new one

jmdelfa committed:
Updated MissingActivityTemplateConflict and MissingAssociationConflict. They incorporate the anchorId (if any) and whether the anchor is to Start as information for PrioritySolver.java to solve the conflict

jmdelfa committed:
CoexistenceGoal.java includes the new logic to create conflicts based on anchors. The logic is:
1. If persistentTimeAnchor is disabled, then no anchors are created. There are two scenarios:
1.1 Matching activity found: MissingAssociationConflict created.
1.2 No matching activity found: MissingActivityTemplateConflict created

2. If persistentTimeAnchor is enabled (START or END) then an anchor will be created. There are three scenarios
2.1 Matching activity with anchor found. MissingAssociationConflict created.
2.2 Matching activity found, but only without anchor. MissingAssociationConflict created, passing the ID of the matching activity in order to create the anchor
2.3 No matching activity found: MissingActivityTemplateConflict created

Mythicaeda commented:
I'm concerned about the e2e test failure:

```
SchedulingTests > schedulingPostsSimResults() FAILED
    java.lang.NullPointerException: Cannot invoke "java.lang.Integer.intValue()" because the return value of "gov.nasa.jpl.***.e2e.types.SchedulingResponse.datasetId()" is null
        at gov.nasa.jpl.***.e2e.SchedulingTests.schedulingPostsSimResults(SchedulingTests.java:358)
```

`datasetId` is only `null` if no simulation results were posted as a result of scheduling. It should not be `null` in this case, as the first time the plan in this test is simulated is within the scheduling run.

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpressionRelativeBinary.java:
```diff
@@ -20,4 +23,9 @@ public Interval computeTime(final SimulationResults simulationResults, final Pla
     return Interval.between(interval1.start, interval2.end);
   }
+
+  public TimeAnchor getAnchor(){
+    return this.lowerBound.anchor != null ? this.lowerBound.anchor : this.upperBound.anchor;
+  }
```
Can you explain what the rationale here ? In `TimeExpressionRelativeBefore` above you return `null`.

jmdelfa replied:
In a binary relation, both lower and upper bound are TimeExpressionRelativeSimple, each with a TimeAnchor. By default the lowerbound is the one returned

jmdelfa replied:
As discussed, removed and returning empty

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpressionRelativeBefore.java:
```diff
@@ -31,4 +31,9 @@ public Interval computeTime(final SimulationResults simulationResults, final Pla
+
+    @Override
+    public TimeAnchor getAnchor() {
+      return null;
```
Callers of `getAnchor` are protected from `NullPointerException` ? Otherwise, the fact you have to return `null` here is maybe an indication that `getAnchor` as an abstract method is not the right design? Maybe return `Optional<TimeAnchor>` instead ?

jmdelfa replied:
I have updated the code adding Optional<TimeAnchor>. I think that's the best approach

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpressionRelative.java:
```diff
@@ -14,121 +13,90 @@
-  public static TimeExpression afterStart() {
-    TimeExpression te = new TimeExpressionRelativeFixed(TimeAnchor.START, false);
+  public static TimeExpressionRelative afterStart() {
+    TimeExpressionRelative te = new TimeExpressionRelativeSimple(TimeAnchor.START, false);
```
Not blocking: you could have `final var` all of these statements creating new objects in these static methods.

jmdelfa replied:
right. In this case, TimeExpressionRelative is a refactoring of TimeExpression. All these methodds with a specific type are as they were in develop, reason why I did not modified them

jmdelfa replied:
I have corrected it anyways.

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/RecurrenceGoal.java:
```diff
@@ -217,12 +221,15 @@ private java.util.Collection<MissingActivityConflict> makeRecurrenceConflicts(Du
 
     for (var intervalT = start.plus(recurrenceInterval.max);
-         ;
+        ;
```
space has been mistakenly removed

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/RecurrenceGoal.java:
```diff
@@ -217,12 +221,15 @@ private java.util.Collection<MissingActivityConflict> makeRecurrenceConflicts(Du
-        conflicts.add(new MissingActivityTemplateConflict(this, windows, this.getActTemplate(), evaluationEnvironment, 1, Optional.empty()));
+        conflicts.add(new MissingActivityTemplateConflict(this, windows, this.getActTemplate(), evaluationEnvironment, 1,Optional.empty(), Optional.empty(), Optional.empty()));
+      }
+      else{
+        System.out.println();
       }
```
remove this print and the `else` block

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/SimulationData.java:
```diff
@@ -1,11 +1,16 @@
     SimulationResults driverResults,
-    gov.nasa.jpl.aerie.constraints.model.SimulationResults constraintsResults
+    gov.nasa.jpl.aerie.constraints.model.SimulationResults constraintsResults,
+
```
remove useless newline

jmdelfa replied:
done

adrienmaillard commented on scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java:
```diff
@@ -229,14 +231,17 @@ public void schedule(
         //TODO: make revision confirmation atomic part of plan mutation (plan might have been modified during scheduling!)
         ensurePlanRevisionMatch(specification, getMerlinPlanRev(specification.planId()));
-        final var instancesToIds = storeFinalPlan(
+        var instancesToIds = storeFinalPlan(
```
why remove the `final` ?

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java:
```diff
@@ -281,6 +286,26 @@ public void schedule(
 
+  public List<SchedulingActivityDirective> updateEverythingWithNewAnchorIds(Plan solutionPlan, Map<SchedulingActivityDirective, ActivityDirectiveId> instancesToIds){
+    ArrayList<SchedulingActivityDirective> updatedActs = new ArrayList<SchedulingActivityDirective>();
+
```
this newline and the newline below seem superfluous

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/SchedulingActivityDirective.java:
```diff
@@ -156,6 +176,11 @@ public static SchedulingActivityDirective of(SchedulingActivityDirective o) {
     );
   }
 
+  @Override
+  public ActivityType type() {
+    return type;
+  }
+
```
As `SchedulingActivityDirective` is a record, it already has an built-in accessor for type, no ?

jmdelfa replied:
Indeed. I believe that was legacy code. Removed

adrienmaillard replied:
how come it appears as new code here ?

adrienmaillard commented on scheduler-worker/scheduling-dsl-compiler/src/libs/scheduler-edsl-fluent-api.ts:
```diff
@@ -817,15 +828,16 @@ declare global {
         S extends WindowsEDSL.Gen.ActivityType,
         B extends WindowsEDSL.Gen.ActivityType>(opts: ({
       activityTemplate: (( interval: WindowsEDSL.Interval ) => ActivityTemplate<S>) | ActivityTemplate<S>,
-      forEach:  WindowsEDSL.Windows,
```
Where does this change comes from ? it does not seem to be part of the scope. Was this not there before your work ? :confused:

jmdelfa replied:
As discussed, the global representation of CoexistenceGoal was missing some expressions available for the local representation in the foreach block, namely the WindowsEDSL.Interval and Temporal.Instant

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/activities/ActivityExpression.java:
```diff
@@ -348,7 +348,66 @@ public boolean matches(
+      final @NotNull SchedulingActivityDirective act,
+      final SimulationResults simulationResults,
+      final EvaluationEnvironment evaluationEnvironment,
+      final boolean matchArgumentsExactly, Plan plan) {
+    boolean match = (type == null || type.getName().equals(act.getType().getName()));
```
add a newline and `final` for the `plan` argument

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/model/PlanInMemory.java:
```diff
@@ -198,4 +212,15 @@ public Evaluation getEvaluation() {
+    if(act.anchorId() != null){
+      SchedulingActivityDirective parent = this.getActivitiesById().get(act.anchorId());
+      //todo if parent.duration is null, then
+      return calculateAbsoluteStartOffsetAnchoredActivity(parent).plus(act.anchoredToStart() ? act.startOffset() : act.startOffset().plus(parent.duration()));
```
there is a possibility the parent does not have a duration and it could lead to a failure here if an activity is anchored to its end. Can you add a test and an explicit error message saying that you can't compute the offset if the parent has a `null` duration in this case ?

jmdelfa replied:
Included the following condition in PlanInMemory.calculateAbsoluteStartOffsetAnchoredActivity to capture this behavior:
`      if(!act.anchoredToStart() && parent.duration() == null)
        throw new IllegalArgumentException("Cannot calculate the absolute duration for an activity that is not anchored to the start while the parent doesn't have duration");
`

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/conflicts/MissingAssociationConflict.java:
```diff
@@ -4,27 +4,48 @@
    *
    * @param goal IN STORED the dissatisfied goal that issued the conflict
    * @param instancesToChooseFrom IN the list of instances to choose from to perform the association
+   * @param anchorToStart
```
complete comment

jmdelfa replied:
done

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java:
```diff
@@ -165,8 +175,10 @@ protected CoexistenceGoal fill(CoexistenceGoal goal) {
+
       if(name==null){
-        goal.name = "CoexistenceGoal_forEach_"+forEach.prettyPrint()+"_thereExists_"+this.thereExists.type().getName();
+        goal.name = "CoexistenceGoal_forEach_"+forEach.prettyPrint("")+"_thereExists_"+this.thereExists.type().getName();
```
can you say what adding `""` does ?

jmdelfa replied:
legacy code. removed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java:
```diff
@@ -233,26 +246,34 @@ else if (this.initiallyEvaluatedTemporalContext == null) {
-        Interval startTimeRange = null;
-        startTimeRange = this.startExpr.computeTime(simulationResults, plan, window.interval());
+
+      final var directiveTimeInterval = Interval.between(window.interval().end, Interval.Inclusivity.Inclusive, planHorizon.getEndAerie(), Interval.Inclusivity.Inclusive);
```
maybe rename `directiveTimeInterval` to `afterEndWindow` or something closer to what it is

jmdelfa replied:
done

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java:
```diff
@@ -261,45 +282,89 @@ else if (this.initiallyEvaluatedTemporalContext == null) {
               1,
-              Optional.empty()));
+              anchorValue,
+                  Optional.of(this.persistentAnchor.equals(PersistentTimeAnchor.START)),
```
too much indentation

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java:
```diff
@@ -666,11 +679,43 @@ else if(!analysisOnly &&  (missing instanceof MissingActivityTemplateConflict mi
+                    startOffset
+                );
+                replaceActivity(act,replacementAct);
+                act = replacementAct;
+                //decision-making here, we choose the first satisfying activity
```
why replace `act` by `replacementAct` ? Can't you just use `replacementAct` in the statement below ? as it is the iterator variable for values in actToChooseFrom, I find it odd to assign it.

jmdelfa replied:
agree. removed assignment. replacementAct used directly

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java:
```diff
@@ -705,7 +750,8 @@ private Collection<Conflict> getConflicts(Goal goal) throws SchedulingInterrupte
     synchronizeSimulationWithSchedulerPlan();
     final var evaluationEnvironment = new EvaluationEnvironment(this.problem.getRealExternalProfiles(), this.problem.getDiscreteExternalProfiles());
-    final var rawConflicts = goal.getConflicts(plan, lastSimulationResults, evaluationEnvironment, this.problem.getSchedulerModel());
+    Optional<BidiMap<SchedulingActivityDirectiveId, ActivityDirectiveId>> mapSchedulingIdsToActivityIds = simulationFacade.getBidiActivityIdCorrespondence();
```
could it be final ?

jmdelfa replied:
yes. Fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java:
```diff
@@ -812,11 +858,33 @@ private Collection<SchedulingActivityDirective> getBestNewActivities(MissingActi
-        act.ifPresent(newActs::add);
+        if(act.isPresent()){
+          if (missingTemplate.getAnchorId().isPresent()) {
+            SchedulingActivityDirective predecessor = plan.getActivitiesById().get(missingTemplate.getAnchorId().get());
+            int includePredDuration = 0;
+            if (missingTemplate.getAnchorToStart().isPresent()) {
+              includePredDuration = missingTemplate.getAnchorToStart().get() ? 0 : 1;
+            }
+            Duration dur = predecessor.duration().times(includePredDuration);
+            Duration startOffset = act.get().startOffset().minus(plan.calculateAbsoluteStartOffsetAnchoredActivity(predecessor).plus(dur));
```
`predecessor`, `dur`, `startOffset` could be `final var`

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/solver/PrioritySolver.java:
```diff
@@ -923,31 +991,32 @@ private Windows narrowGlobalConstraints(
   }
   private Optional<SchedulingActivityDirective> instantiateActivity(
-      final ActivityExpression activityExpression,
+      final MissingActivityTemplateConflict missingConflict,
```
why do you pass the conflict if below you are only using the template ?

jmdelfa replied:
Right. Fixed

adrienmaillard commented on e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SchedulingTests.java:
```diff
@@ -208,7 +245,7 @@ void schedulingCoexistenceGoal() throws IOException {
     try {
       // Schedule and get Plan
-      hasura.awaitScheduling(schedulingSpecId);
+      hasura.awaitScheduling(schedulingSpecId, 360);
```
360 ?

jmdelfa replied:
removed

adrienmaillard commented on scheduler-driver/src/test/java/gov/nasa/jpl/aerie/scheduler/PrioritySolverTest.java:
```diff
@@ -252,7 +252,7 @@ public void getNextSolution_coexistenceGoalOnActivityWorks_withInitialSimResults
     adHocFacade.computeSimulationResultsUntil(problem.getPlanningHorizon().getEndAerie());
     final var simResults = adHocFacade.getLatestDriverSimulationResults().get();
-    problem.setInitialPlan(makePlanA012(problem), Optional.of(simResults));
+    problem.setInitialPlan(makePlanA012(problem), Optional.of(simResults), null);
```
can't you get the id map from the `adHocFacade` and pass it ?

jmdelfa replied:
done

adrienmaillard commented on scheduler-driver/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java:
```diff
@@ -1183,6 +1186,65 @@ public void testCoexistenceExternalResource() throws SchedulingInterruptedExcept
 
+  @Test
+  public void testCoexistenceWithAnchors() throws SchedulingInterruptedException {
+    Interval period = Interval.betweenClosedOpen(Duration.of(0, Duration.HOURS), Duration.of(20, Duration.HOURS));
```
period could be `final var`

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/test/java/gov/nasa/jpl/aerie/scheduler/TestApplyWhen.java:
```diff
@@ -1183,6 +1186,65 @@ public void testCoexistenceExternalResource() throws SchedulingInterruptedExcept
+
+    //have some activity already present
+    //  create a PlanInMemory, add ActivityInstances
+    PlanInMemory partialPlan = new PlanInMemory();
```
`partialPlan` could be `final var`

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-worker/src/test/java/gov/nasa/jpl/aerie/scheduler/worker/services/SchedulingDSLCompilationServiceTests.java:
```diff
@@ -655,6 +661,7 @@ export default function() {
                       Map.entry("duration", new ProfileExpression<>(new DiscreteProfileFromDuration(new DurationLiteral(Duration.of(1, HOUR)))))
                   ))
               ),
+                  Optional.of(PersistentTimeAnchor.DISABLED),
```
are we sure about the indentation of these new anchor members ? they seem off.

jmdelfa replied:
Optional.of in line 664 was off. Fixed

adrienmaillard commented on e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SchedulingTests.java:
```diff
@@ -274,11 +274,12 @@ void schedulingCoexistenceGoal() throws IOException {
         modelId,
-        coexistenceGoalWithAnchorsDefinition);
-    hasura.createSchedulingSpecGoal(coexistenceGoalId, schedulingSpecId, 0);
+        coexistenceGoalWithAnchorsDefinition);*/
```
remove this commented out code

jmdelfa replied:
removed

adrienmaillard commented on e2e-tests/src/test/java/gov/nasa/jpl/aerie/e2e/SchedulingTests.java:
```diff
@@ -274,11 +274,12 @@ void schedulingCoexistenceGoal() throws IOException {
   void schedulingCoexistenceGoalWithAnchor() throws IOException {
```
these fixes could be squashed with the other changes you made previously to `SchedulingTests`.

adrienmaillard reviewed (changes requested):
No major problem. Nice testing.

jmdelfa committed:
Updated PrioritySolver.java to handle conflicts with anchors.
The scenarios are:
If existing activity act is a match but is missing the anchor, then the appropriate anchorId has been included in MissingAssociationConflict. In that case, a new activity must be created as a copy of act but including the anchorId. This activity is then added to all appropriate data structures and the association is created
If there is no match, a MissingActivityTemplateConflict.java has been created, including the ID of the directive to which the new activity should be anchored. In that case, we calculate the absolute start offset and create the new activity with the anchorid

jmdelfa committed:
New tests with PersistentTimeAnchor

adrienmaillard commented on scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java:
```diff
@@ -281,6 +286,24 @@ public void schedule(
     }
   }
 
+  public List<SchedulingActivityDirective> updateEverythingWithNewAnchorIds(Plan solutionPlan, Map<SchedulingActivityDirective, ActivityDirectiveId> instancesToIds){
+    ArrayList<SchedulingActivityDirective> updatedActs = new ArrayList<SchedulingActivityDirective>();
+    var planActs = new HashSet<>(solutionPlan.getActivities());
+    for (SchedulingActivityDirective act : planActs) {
+      if (act.anchorId() != null) {
+        SchedulingActivityDirective actAnchored = solutionPlan.getActivitiesById().get(act.anchorId());
+        SchedulingActivityDirective updatedAct = SchedulingActivityDirective.copyOf(act, new SchedulingActivityDirectiveId(instancesToIds.get(actAnchored).id()), act.anchoredToStart(), act.startOffset());
+        updatedActs.add(updatedAct);
+        solutionPlan.replaceActivity(act, updatedAct);
+        ActivityDirectiveId value = instancesToIds.get(act);
+        instancesToIds.remove(act);
+        instancesToIds.put(updatedAct, value);
+      }
+    }
+    return updatedActs;
+  }
+
+
```
do a pass for `final`

jmdelfa replied:
Done. All local vars were converted to final var

adrienmaillard commented on scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java:
```diff
@@ -390,6 +413,7 @@ private PlanComponents loadInitialPlan(
       final Optional<SimulationResults> initialSimulationResults) {
     //TODO: maybe paranoid check if plan rev has changed since original metadata?
     try {
+      BidiMap<SchedulingActivityDirectiveId, ActivityDirectiveId> mapSchedulingIdsToActivityIds =  new DualHashBidiMap<SchedulingActivityDirectiveId, ActivityDirectiveId>();
```
this could be final

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/test/java/gov/nasa/jpl/aerie/scheduler/TestPersistentAnchor.java:
```diff
@@ -0,0 +1,587 @@
+  /* Test cases in which the goal is created with TimeExpression endsAt
+   */
+
+  public TestData createTestCaseStartsAt(PersistentTimeAnchor persistentAnchor,  boolean missingActAssociationsWithAnchor, boolean missingActAssociationsWithoutAnchor, int activityDurationHours, int goalStartPeriodHours, int goalEndPeriodHours, TimeAnchor timeAnchor, TimeExpressionRelative timeExpression, long relativeOffsetHours)
+  throws SchedulingInterruptedException
```
all of these parameters should be `final`. Please do a pass over this file for final modifier, I saw other occurrences.

jmdelfa replied:
all TestPersistentAnchor reviewed/updated with (final) var when applicable

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/goals/CoexistenceGoal.java:
```diff
@@ -260,45 +282,91 @@ else if (this.initiallyEvaluatedTemporalContext == null) {
+          ));
         } else {
-          conflicts.add(new MissingAssociationConflict(this, missingActAssociations));
+            var actsToAssociate = missingActAssociationsWithAnchor.isEmpty() ? missingActAssociationsWithoutAnchor : missingActAssociationsWithAnchor;
```
this could be final

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpressionRelativeBinary.java:
```diff
@@ -4,12 +4,14 @@
+  private final TimeExpressionRelativeSimple lowerBound;
+  private final TimeExpressionRelativeSimple upperBound;
+
+  public TimeExpressionRelativeBinary(TimeExpressionRelativeSimple lowerBound, TimeExpressionRelativeSimple upperBound) {
```
parameters should be final

jmdelfa replied:
fixed

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/constraints/timeexpressions/TimeExpressionRelative.java:
```diff
@@ -0,0 +1,150 @@
+      return this;
+    }
+
+
+    public Builder name(String name) {
+      this.name = name;
+      return getThis();
+    }
+
+    public Builder from(TimeExpressionRelative otherExpr) {
+      fromExpression = otherExpr;
+      return getThis();
+    }
+
+    TimeExpressionRelative fromExpression;
+
+    public Builder from(TimeAnchor anchor) {
+      fromAnchor = anchor;
+      return getThis();
+    }
+
+    TimeAnchor fromAnchor;
+
+    public Builder minus(Duration dur) {
+      operations.add(Pair.of(TimeUtility.Operator.MINUS, dur));
+      return getThis();
+    }
+
+    public Builder interval() {
+      this.interval = true;
+      return getThis();
+    }
+
+    public Builder plus(Duration dur) {
+      operations.add(Pair.of(TimeUtility.Operator.PLUS, dur));
+      return getThis();
+    }
```
all function parameters should be final

adrienmaillard commented on scheduler-driver/src/main/java/gov/nasa/jpl/aerie/scheduler/simulation/SimulationFacade.java:
```diff
@@ -146,20 +149,31 @@ public void setActivityTypes(final Collection<ActivityType> activityTypes){
   }
 
   public Map<SchedulingActivityDirectiveId, ActivityDirectiveId> getActivityIdCorrespondence(){
-    return planActDirectiveIdToSimulationActivityDirectiveId;
+    return new HashMap<>(mapSchedulingIdsToActivityIds);
+  }
+
+  public Optional<BidiMap<SchedulingActivityDirectiveId, ActivityDirectiveId>> getBidiActivityIdCorrespondence(){
+    if(initialSimulationResults.isEmpty() || initialPlanHasBeenModified)
+      return Optional.ofNullable(mapSchedulingIdsToActivityIds);
+    else if(initialSimulationResults.isPresent())
+      return initialSimulationResults.get().mapSchedulingIdsToActivityIds();
+    else
+      return Optional.empty();
   }
```
why keep both maps ? is `getActivityIdCorrespondence()` still used ?

jmdelfa replied:
The coversion from bidi to hashmap is done as it is used in SynchronousSchedulerAgent:
```java
    final var schedID_to_simID =
        simulationFacade.getActivityIdCorrespondence();
```

adrienmaillard replied:
I would move the conversion (and thus delete `getActivityIdCorrespondence`) out of this class as the logic you added in getBidiAct... is not present in `getActivityIdCorrespondence`.

adrienmaillard reviewed (changes requested):
We're almost there. I'll approve after this last batch.

jmdelfa committed:
v0.1-msr dev. Unstable. Implementing PEL, system and subsystem generic classes and mechanism based on EnumMap to define hierarchically subsystem states for system states

jmdelfa committed:
SimulationFacade had a method getActivityIdCorrespondence which returned a HashMap. That has been superseded by getBidiAcitivbityIdCorrespondence which uses the new type BidiMap. The old method has been removed and the only call to the getActivityIdCorrespondence, in class SynchronousSchedulerAgent, has been modified to receive a BidiMap and then convert to HashMap so that the existing logic doesn't need to be changed.

adrienmaillard reviewed (approved):
Looks good now ! thank you.

dandelany commented:
Thanks! This had a failing test before but it passed on re-run so I think it was a fluke. Rebased & merging now.
